src/index.js

```diff
@@ -12,15 +14,32 @@ const Heartbeat = require('./heartbeat')
 
 class GossipSub extends BasicPubsub {
   /**
-   * @param {Object} libp2p an instance of Libp2p
-   * @param {Object} options
-   * @param {bool} options.emitSelf if publish should emit to self, if subscribed, defaults to false
-   * @param {bool} options.gossipIncoming if incoming messages on a subscribed topic should be automatically gossiped, defaults to true
-   * @param {bool} options.fallbackToFloodsub if dial should fallback to floodsub, defaults to true
+   * @param {PeerInfo} peerInfo instance of the peer's PeerInfo
+   * @param {Object} registrar
+   * @param {function} registrar.handle
+   * @param {function} registrar.register
+   * @param {function} registrar.unregister
+   * @param {Object} [options]
+   * @param {bool} [options.emitSelf] if publish should emit to self, if subscribed, defaults to false
+   * @param {bool} [options.gossipIncoming] if incoming messages on a subscribed topic should be automatically gossiped, defaults to true
+   * @param {bool} [options.fallbackToFloodsub] if dial should fallback to floodsub, defaults to true
    * @constructor
    */
-  constructor (libp2p, options) {
-    super('libp2p:gossipsub', constants.GossipSubID, libp2p, options)
+  constructor (peerInfo, registrar, options = {}) {
+    assert(PeerInfo.isPeerInfo(peerInfo), 'peer info must be an instance of `peer-info`')
+
+    // registrar handling
+    assert(registrar && typeof registrar.handle === 'function', 'a handle function must be provided in registrar')
+    assert(registrar && typeof registrar.register === 'function', 'a register function must be provided in registrar')
+    assert(registrar && typeof registrar.unregister === 'function', 'a unregister function must be provided in registrar')
```

Comment:
Can these three checks be pulled into some sort of utility function, `similarly to PeerInfo.isPeerInfo`?
I suspect there are other users of `registrar`s that could use it.

Reply:
I will just have this in `libp2p-pubsub` for now. After that, I may add `class-is` to `Registrar`